Fix bug in ControlNetPipelines with MultiControlNetModel of length 1

greentfrapp opened the pull request:
# What does this PR do?

Hey @patrickvonplaten and @sayakpaul!

In the latest implementation that added `control_guidance_start` and `control_guidance_end`, the code that computes `controlnet_keep` uses `len(keeps) == 1` to decide whether to `controlnet_keep` should be a list of floats or list of lists.

But ControlNetPipelines supplied with a **list of one ControlNetModel** initializes a MultiControlNetModel as well.

This causes a bug when a ControlNetPipeline is supplied with a **list of one ControlNetModel** because that initializes a MultiControlNetModel but `controlnet_keep` is computed as a list of floats.
On runtime an error will be raised when computing `cond_scale` at each timestep - `TypeError: can't multiply sequence by non-int of type 'float'`.

This fix changes the following in the 3 ControlNetPipeline files:

Previous code:
```
controlnet_keep.append(keeps[0] if len(keeps) == 1 else keeps)
```

Fixed code:
```
controlnet_keep.append(keeps[0] if isinstance(controlnet, ControlNetModel) else keeps)
```

The fix accounts for the edge case where MultiControlNetModels have only one controlnet.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_4032). All of your documentation changes will be reflected on that endpoint.

sayakpaul reviewed (approved):
Thanks for the fix!

Can we maybe add some tests to ensure the robustness of the changes? 

Also cc @holwech.

greentfrapp commented:
Sure thing - I can help to add some tests for this after my work today, within the next 12 hours

greentfrapp commented:
@sayakpaul I've updated a few of the tests in `tests/pipelines/controlnet/test_controlnet.py` to test with varying number of ControlNet models as well.

Let me know if that's sufficient?

Not sure if it's necessary to also add the same for `test_controlnet_img2img.py` and `test_controlnet_inpaint.py`.
Also, let me know if there are issues with the way I've implemented the updated tests.

sayakpaul reviewed (commented):
Thanks for your changes. 

I think we should keep the existing tests untouched. I'd rather just write separate test cases (even if they seem redundant) to just test this behavior. Since we call these "Fast" tests, it's okay if we are being redundant here. 

Let me know if I am missing out on something.

greentfrapp commented:
@sayakpaul Sure thing, I've reverted the existing tests and created a new set of tests for this behavior - essentially duplicating the tests for MultiControlNet and setting number of models to a list of 1. Let me know if I missed out anything?

sayakpaul commented on tests/pipelines/controlnet/test_controlnet.py:
```diff
@@ -398,6 +398,179 @@ def test_save_pretrained_raise_not_implemented_exception(self):
                 pass
 
 
+class StableDiffusionMultiControlNetOneModelPipelineFastTests(
```
This works for me. Thanks much!

sayakpaul reviewed (approved):
Thanks so much for iterating!

@patrickvonplaten @yiyixuxu could you also give this a look?

patrickvonplaten reviewed (approved):
Great catch! Thanks for the cool PR

yiyixuxu reviewed (approved):
Awesome thank you! 

Is `MultiControlnetModel` with one model work the same as a `ControlnetModel`? If so maybe (in a future PR) we can simplify the code by only keeping the `MultiControlnetModel` concept?

patrickvonplaten commented:
> Awesome thank you!
> 
> Is `MultiControlnetModel` with one model work the same as a `ControlnetModel`? If so maybe (in a future PR) we can simplify the code by only keeping the `MultiControlnetModel` concept?

Actually yes I think it would make sense to only use `MultiControlnetModel` in the code

greentfrapp commented:
Should this PR be merged for now? Or do we want to refactor the ControlNet code to use only MultiControlnet? Might also have to update some documentation.

sayakpaul commented:
I think this PR should be merged now. In a future PR, as @yiyixuxu suggested we could unify the MultiControlNetModel logic which should simplify the code a bit. 

@patrickvonplaten @yiyixuxu okay for you if we merged the PR now?
